dhall/src/Dhall/Schemas.hs

```diff
@@ -0,0 +1,206 @@
+
+    let schemasText = Pretty.Text.renderStrict docStream
+
+    case outputMode of
+        Write -> do
+            case input of
+                InputFile file -> do
+                    if originalText == schemasText
+                        then return ()
+                        else AtomicWrite.atomicWriteFile
+                                file
+                                (Pretty.Text.renderLazy docStream)
+                StandardInput -> do
+                    supportsANSI <- ANSI.hSupportsANSI IO.stdout
+
+                    Pretty.Terminal.renderIO
+                        IO.stdout
+                        (if supportsANSI
+                            then fmap Dhall.Pretty.annToAnsiStyle docStream
+                            else Pretty.unAnnotateS docStream)
+
+        Check -> do
+            if originalText == schemasText
+                then return ()
+                else do
+                    let command = "rewrite-with-schemas"
+
+                    let modified = "rewritten"
+
+                    Exception.throwIO CheckFailed{..}
+
+decodeSchema :: Expr Src Void -> Maybe (Expr Src Void, Map Text (Expr Src Void))
+decodeSchema (RecordLit m)
+        | Just  _Type               <- Map.lookup "Type" m
+        , Just (RecordLit _default) <- Map.lookup "default" m =
+            Just (_Type, _default)
+decodeSchema _ =
+    Nothing
+
+decodeSchemas
+    :: Expr Src Void
+    -> Maybe (Map SHA256Digest (Text, Map Text (Expr Src Void)))
+decodeSchemas (RecordLit keyValues) = do
+    m <- traverse decodeSchema keyValues
+
+    let typeMetadata = Map.fromList $ do
+            (name, (_Type, _default)) <- Map.toList m
+
+            return (Import.hashExpression (Syntax.denote _Type), (name, _default))
+
+    return typeMetadata
+decodeSchemas  _ = do
+    empty
+
+-- | Simplify a Dhall expression using a record of schemas
+simplifyUsingSchemas
+    :: Expr Src Import
+    -- ^ Record of schemas
+    -> Expr Src Import
+    -- ^ Expression to simplify using the supplied schemas
+    -> IO (Expr Src Import)
+simplifyUsingSchemas _schemas expression = do
+    resolvedSchemas    <- Import.load _schemas
+    resolvedExpression <- Import.load expression
+
+    _ <- Core.throws (TypeCheck.typeOf resolvedSchemas)
+    _ <- Core.throws (TypeCheck.typeOf resolvedExpression)
+
+    let normalizedSchemas    = Normalize.normalize resolvedSchemas
+    let normalizedExpression = Normalize.normalize resolvedExpression
+
+    typeMetadata <- case decodeSchemas normalizedSchemas of
+        Just typeMetadata -> return typeMetadata
+        Nothing           -> Exception.throwIO NotASchemaRecord
+
+    let schemasRewrite subExpression@(RecordLit keyValues) =
+            Maybe.fromMaybe subExpression $ do
+                let substitutions = Map.singleton "schemas" normalizedSchemas
+
+                let substitutedExpression =
+                        Substitution.substitute (Syntax.denote subExpression) substitutions
```

Comment:
Since `subExpression` is already normalized, the `denote` seems unnecessary.

```suggestion
                        Substitution.substitute subExpression substitutions
```

If we need to make the types match up, better `renote` where necessary.